src/main/kotlin/io/sc3/peripherals/client/item/PosterRenderer.kt

```diff
@@ -0,0 +1,245 @@
+package io.sc3.peripherals.client.item
+
+import io.sc3.peripherals.Registration.ModItems
+import io.sc3.peripherals.ScPeripherals.ModId
+import io.sc3.peripherals.posters.PosterItem
+import io.sc3.peripherals.posters.PosterState
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
+import net.fabricmc.api.EnvType
+import net.fabricmc.api.Environment
+import net.minecraft.block.MapColor
+import net.minecraft.client.MinecraftClient
+import net.minecraft.client.render.RenderLayer
+import net.minecraft.client.render.VertexConsumer
+import net.minecraft.client.render.VertexConsumerProvider
+import net.minecraft.client.render.item.HeldItemRenderer
+import net.minecraft.client.texture.NativeImageBackedTexture
+import net.minecraft.client.util.math.MatrixStack
+import net.minecraft.entity.EntityType
+import net.minecraft.entity.decoration.ItemFrameEntity
+import net.minecraft.item.ItemStack
+import net.minecraft.util.math.RotationAxis
+import org.joml.Matrix4f
+import java.util.function.BiFunction
+
+@Environment(EnvType.CLIENT)
+object PosterRenderer : AutoCloseable {
+  val POSTER_BACKGROUND_RES = ModId("textures/item/poster_background.png")
+  val POSTER_BACKGROUND = RenderLayer.getText(POSTER_BACKGROUND_RES)
+
+//  val MAP_ICONS_RENDER_LAYER = RenderLayer.getText(MAP_ICONS_TEXTURE)
+//  private const val DEFAULT_IMAGE_WIDTH = 128
+//  private const val DEFAULT_IMAGE_HEIGHT = 128
+
+  private val textureManager get() = MinecraftClient.getInstance().textureManager
+
+  private val posterTextures = mutableMapOf<String, PosterTexture>() // : Int2ObjectMap<PosterTexture> = Int2ObjectOpenHashMap()
+  fun updateTexture(id: String, state: PosterState) {
+    getPosterTexture(id, state).setNeedsUpdate()
+  }
+
+  fun draw(
+    matrices: MatrixStack,
+    vertexConsumers: VertexConsumerProvider,
+    id: String,
+    state: PosterState,
+    light: Int
+  ) {
+    getPosterTexture(id, state).draw(matrices, vertexConsumers, light)
+  }
+
+  fun drawCropped(
+    matrices: MatrixStack,
+    vertexConsumers: VertexConsumerProvider,
+    id: String,
+    state: PosterState,
+    light: Int,
+    cropX: Float,
+    cropY: Float,
+    cropWidth: Float,
+    cropHeight: Float
+  ) {
+    getPosterTexture(id, state).drawCropped(matrices, vertexConsumers, light, cropX, cropY, cropWidth, cropHeight)
+  }
+
+  private fun getPosterTexture(id: String, state: PosterState): PosterTexture {
+    return posterTextures.compute(id,
+      BiFunction<String, PosterTexture?, PosterTexture?> { id2: String, texture: PosterTexture? ->
+        if (texture == null) {
+          return@BiFunction PosterTexture(id2, state)
+        } else {
+          texture.setState(state)
+          return@BiFunction texture
+        }
+      })!!
```

Comment:
Can this be changed to something like:

```suggestion
    return posterTextures.computeIfAbsent(id) { PosterTexture(id, state) }
      .also { it.setState(state) }
```

It's ok if you think the current code is better. getOrPut would be fine instead of computeIfAbsent too. setState() would get called twice for a new poster here but from what I can tell it doesn't need to be atomic and setState does an equality check to decide if needsUpdate needs to be set anyway.